internal/pkg/earpointer/taint.go

```diff
@@ -153,31 +155,34 @@ func (ht *heapTraversal) canReach(sink ssa.Instruction, sources []*source.Source
 	return nil
 }
 
-// For a function, transitively get the functions called within this function.
-// Argument "depth" controls the depth of the call chain.
-// For example, return {g1,g2,g3} for "func f(){ g1(); g2() }, func g1(){ g3() }".
-func calleeFunctions(fn *ssa.Function, result map[*ssa.Function]bool, depth uint) {
-	if depth <= 0 {
+// For a function, transitively get the functions reachable from this function
+// according to the call graph. Both callers and callees are considered.
+// Argument "depth" controls the depth of the call chain. For example,
+//   func f(){ g1(); g2() }
+//   func g1(){ g3() }
+// for input "g1", return {f,g1,g2,g3} if depth>1, and return {f, g1} if depth=1.
+func reachableFunctions(fn *ssa.Function, result map[*ssa.Function]bool, cg *callgraph.Graph, depth uint) {
+	if depth <= 0 || result[fn] {
 		return
 	}
-	for _, b := range fn.Blocks {
-		for _, instr := range b.Instrs {
-			if call, ok := instr.(*ssa.Call); ok {
-				// TODO(#317): use more advanced call graph.
-				// skip empty, unlinked, or visited functions
-				if callee := call.Call.StaticCallee(); callee != nil && len(callee.Blocks) > 0 && !result[callee] {
-					result[callee] = true
-					calleeFunctions(callee, result, depth-1)
-				}
-			}
-		}
+	result[fn] = true
+	node := cg.Nodes[fn]
+	if node == nil {
+		return
+	}
+	// Visit the callees within "fn".
+	for _, out := range node.Out {
+		reachableFunctions(out.Callee.Func, result, cg, depth-1)
+	}
+	// Visit the callers of "fn".
```

Comment:
What are the implications of also gathering the callers? Does this lead to more true/false positives?

Reply:
It is to capture more inter-procedural taint propagation. The new added tests contain some example. 
For instance, in the following code,`createSource2()` return type `interface{}` rather than  `Source`, so Levee may not report a finding at `Sinkf(createSource2())` since it loses the source type information upon the exit of `createSource2()`.

func createSource2() interface{} {
	return &core.Source{}
}

Sinkf(createSource2())

In general, the goal is to connect the sources and sinks even there are no explicit embedding structure in the functions containing the sources and sinks.

I have seen some new findings in k8s by using this extension, but most of they are likely FPs, which can be partially reduced using context sensitiveness (in the next CL).

Reply:
Sounds good, makes sense to me!